### ISSUE TITLE
Mechanism to enable __weakref__ slot on dataclass(slots=True)

### ISSUE THREAD

**Issue by abbe7814-fbdd-452d-9e9a-6b795243a34a** [python/cpython#91215]
BPO | [47059](https://bugs.python.org/issue47059)
--- | :---
Nosy | @ericvsmith, @ariebovenberg

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2022-03-18.18:00:50.175>
labels = ['library', '3.11']
title = 'Mechanism to enable __weakref__ slot on dataclass(slots=True)'
updated_at = <Date 2022-03-18.18:00:50.175>
user = 'https://github.com/ariebovenberg'
```

bugs.python.org fields:
```python
activity = <Date 2022-03-18.18:00:50.175>
actor = 'ariebovenberg'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2022-03-18.18:00:50.175>
creator = 'ariebovenberg'
dependencies = []
files = []
hgrepos = []
issue_num = 47059
keywords = []
message_count = 1.0
messages = ['415515']
nosy_count = 2.0
nosy_names = ['eric.smith', 'ariebovenberg']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue47059'
versions = ['Python 3.11']
```

</p></details>

**Comment by rhettinger** [comment]
A weakref slot should be opt-in.  No need to make everyone pay for a feature that is so rarely needed.

**Comment by ericvsmith** [comment]
I agree. I would rather this be `weakref_slot=True`, which defaults to `False`.

**Comment by ericvsmith** [comment]
PR #92160

**Comment by ericvsmith** [comment]
#92160 was merged to 3.11.
